### PR TITLE
PIN-2422: a producer of itself must be enabled to work on purposes as a consumer

### PR DESCRIPTION
**Pull request by beetlecrunch**

### CODE REVIEW

File: src/main/scala/it/pagopa/interop/purposeprocess/api/impl/OrganizationRole.scala
Comment (beetlecrunch) [inline src/main/scala/it/pagopa/interop/purposeprocess/api/impl/OrganizationRole.scala]: Any better name is welcome.

File: src/main/scala/it/pagopa/interop/purposeprocess/api/impl/PurposeApiServiceImpl.scala
Comment (beetlecrunch) [inline src/main/scala/it/pagopa/interop/purposeprocess/api/impl/PurposeApiServiceImpl.scala]: In the case of Consumer == Producer, this check will always return a ChangedBy.CONSUMER.
What about replacing it with an approach related to OrganizationRole?

File: src/main/scala/it/pagopa/interop/purposeprocess/api/impl/OrganizationRole.scala
Comment (beetlecrunch) [inline src/main/scala/it/pagopa/interop/purposeprocess/api/impl/OrganizationRole.scala]: I'll be honest, I dislike `PRODUCER_OF_ITSELF`, but no better idea except `CONSUMER_PRODUCER`.
Any suggestions?

File: src/main/scala/it/pagopa/interop/purposeprocess/api/impl/PurposeApiServiceImpl.scala
Comment (TonioGela) [inline src/main/scala/it/pagopa/interop/purposeprocess/api/impl/PurposeApiServiceImpl.scala]: The external for comprehension clutters a lot IMHO, how about a `.map(_.flatten)` on the `traverse` output?

File: src/main/scala/it/pagopa/interop/purposeprocess/api/impl/PurposeVersionActivation.scala
Comment (galales) [inline src/main/scala/it/pagopa/interop/purposeprocess/api/impl/PurposeVersionActivation.scala:100]: too much time passed and I honestly don't remember what we agreed: is it expected that the activation in case of `PRODUCER_OF_ITSELF` should wait for approval in case of exceeded load?

Plus: in case of `(SUSPENDED, PRODUCER | PRODUCER_OF_ITSELF)` the load check is not performed. I guess this check should be done in both cases or neither of them.

File: src/main/scala/it/pagopa/interop/purposeprocess/api/impl/PurposeApiServiceImpl.scala
Comment (TonioGela) [reply src/main/scala/it/pagopa/interop/purposeprocess/api/impl/PurposeApiServiceImpl.scala:256]: here's my proposal
```suggestion
      Future
        .traverse(purposes.purposes)(purpose =>
          for {
            changedBy       <- getChangedBy(organizationId, purpose.eserviceId, purpose.consumerId).attempt
            enhancedPurpose <- changedBy.traverse(enhancePurpose(purpose, _))
          } yield enhancedPurpose.toOption
        )
        .map(_.flatten)
```

File: src/main/scala/it/pagopa/interop/purposeprocess/api/impl/OrganizationRole.scala
Comment (galales) [reply src/main/scala/it/pagopa/interop/purposeprocess/api/impl/OrganizationRole.scala]: An alternative could be `SELF_CONSUMER`, but I'm not so sure that it's much better 😅

File: src/main/scala/it/pagopa/interop/purposeprocess/api/impl/OrganizationRole.scala
Comment (galales) [inline src/main/scala/it/pagopa/interop/purposeprocess/api/impl/OrganizationRole.scala]: may we return an `Either` here?

File: src/main/scala/it/pagopa/interop/purposeprocess/api/impl/PurposeApiServiceImpl.scala
Comment (galales) [reply src/main/scala/it/pagopa/interop/purposeprocess/api/impl/PurposeApiServiceImpl.scala]: I agree, we should use the new logic (also in `assert*` methods)

File: src/main/scala/it/pagopa/interop/purposeprocess/api/impl/OrganizationRole.scala
Comment (galales) [reply src/main/scala/it/pagopa/interop/purposeprocess/api/impl/OrganizationRole.scala]: I agree on changing the name, `OrganizationRole` has a specific and different meaning for us.
Can `Ownership` be a viable alternative?

File: src/main/scala/it/pagopa/interop/purposeprocess/api/impl/PurposeVersionActivation.scala
Comment (beetlecrunch) [reply src/main/scala/it/pagopa/interop/purposeprocess/api/impl/PurposeVersionActivation.scala:100]: Applying the same behavior used for Consumer and Producer.

**Review by TonioGela — APPROVED** [review]

**Review by galales — APPROVED** [review]

**Pull request merged**
